File: C++/2-Medium/00314._二叉树的垂直遍历/00314._二叉树的垂直遍历.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <tuple>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// 函数用于执行二叉树的垂直遍历
vector<vector<int>> verticalOrder(TreeNode* root) {
    if (!root) return {};
    
    // 用来存储结果，外层map的key是列索引，内层map的key是行索引
    map<int, map<int, vector<int>>> nodes;
    // 队列用于BFS，存储节点以及其对应的行和列索引
    queue<tuple<TreeNode*, int, int>> q;
    q.push(make_tuple(root, 0, 0)); // 根节点位于(0, 0)
    
    while (!q.empty()) {
        auto [node, row, col] = q.front();
        q.pop();
        nodes[col][row].push_back(node->val);
        if (node->left) {
            q.push(make_tuple(node->left, row + 1, col - 1)); // 左子节点行+1，列-1
        }
        if (node->right) {
            q.push(make_tuple(node->right, row + 1, col + 1)); // 右子节点行+1，列+1
        }
    }
    
    // 将map中的结果转换为题目要求的格式
    vector<vector<int>> result;
    for (auto& p : nodes) {
        vector<int> colVals;
        for (auto& q : p.second) {
            colVals.insert(colVals.end(), q.second.begin(), q.second.end());
        }
        result.push_back(colVals);
    }
    
    return result;
}

// 主函数用于测试
int main() {
    // 构建测试用例的二叉树
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    
    // 执行垂直遍历
    vector<vector<int>> result = verticalOrder(root);
    // 打印结果
    for (auto& col : result) {
        for (int val : col) {
            cout << val << " ";
        }
        cout << endl;
    }
    
    return 0;
}

File: Python/2-Medium/00314._二叉树的垂直遍历/00314._二叉树的垂直遍历.py
from typing import Optional, List, Tuple, Dict
from collections import deque

class TreeNode:
    """定义二叉树节点"""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def verticalOrder(root: Optional[TreeNode]) -> List[List[int]]:
    """
    二叉树的垂直遍历

    Args:
        root: 二叉树的根节点

    Returns:
        按垂直顺序遍历的节点值列表

    说明:
        使用BFS遍历树，并记录每个节点的位置（列号和行号）。
        根据列号聚合节点值，并根据行号对每列的节点值排序。
    """
    if not root:
        return []

    # 初始化列号与节点值及行号的映射
    columns = {}  # type: Dict[int, List[Tuple[int, int]]]
    queue = deque([(root, 0, 0)])  # 存储（节点，列号，行号）

    while queue:
        node, col, row = queue.popleft()
        
        # 将节点值及其行号添加到对应列号的列表中
        if col not in columns:
            columns[col] = []
        columns[col].append((row, node.val))
        
        # 添加子节点到队列
        if node.left:
            queue.append((node.left, col - 1, row + 1))
        if node.right:
            queue.append((node.right, col + 1, row + 1))

    # 按列号排序，并对每列的节点按行号（和节点值）排序
    sorted_columns = sorted(columns.items())
    result = [[val for _, val in sorted(values)] for _, values in sorted_columns]

    return result

# 构建示例二叉树
root = TreeNode(3)
root.left = TreeNode(9)
root.right = TreeNode(20)
root.right.left = TreeNode(15)
root.right.right = TreeNode(7)

# 运行代码
verticalOrder(root)
